### PR TITLE
Update statement and transaction analytics

### PR DESCRIPTION
**Pull request by SteveOye**

### CODE REVIEW

File: android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt [hunk 1]
```diff
@@ -142,7 +143,60 @@ class FlutterPericulumPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
             }
 
 
-        } else if (call.method == "generateMobileDataAnalysis") {
+        } else if (call.method == "getStatementAnalytics"){
+            println("getStatementAnalytics..")
```
Comment (zabeehpericulum) [inline android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt:147]: probably best to remove print statements

File: android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt [hunk 2]
```diff
@@ -142,7 +143,60 @@ class FlutterPericulumPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
...
+                        override fun onFailure(call: Call, e: IOException) {
+                            val error = e.message
+                            result.success("{\"title\": \"${error}\"}")
+                        }
+                    })
+
+            
+           }
+       } else if (call.method == "getStatementTransaction"){
+            println("getStatementTransaction..")
```
Comment (zabeehpericulum) [inline android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt:173]: no println please

File: android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt [hunk 3]
```diff
@@ -142,7 +143,60 @@ class FlutterPericulumPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
             }
 
 
-        } else if (call.method == "generateMobileDataAnalysis") {
+        } else if (call.method == "getStatementAnalytics"){
+            println("getStatementAnalytics..")
+            val args = call.arguments as HashMap<String, Any>
+            if (args != null){
+                val token = args.get("token")
+                val endpoint = "statement"
```
Comment (zabeehpericulum) [inline android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt:151]: I don't think you used this.

File: example/lib/main.dart [hunk 4]
```diff
@@ -29,6 +31,7 @@ class _MyAppState extends State<MyApp> {
 
   @override
   Widget build(BuildContext context) {
+    const token = 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1VSkJOVUk0UkRFek9FVTBORGd4UWpVMVJqTTJPVEJEUXpRMFF6bEJRa1F6UWpnd1JETkVSQSJ9.eyJodHRwczovL2luc2lnaHRzLXBlcmljdWx1bS5jb20vdGVuYW50IjoibnVjbGV1c2lzIiwiaXNzIjoiaHR0cHM6Ly9wZXJpY3VsdW0tdGVjaG5vbG9naWVzLWluYy5hdXRoMC5jb20vIiwic3ViIjoiSDR1VHJzdjJoMGlEVGlTMDR2NmVGWmNpdTNLMGJvWnJAY2xpZW50cyIsImF1ZCI6Imh0dHBzOi8vYXBpLmluc2lnaHRzLXBlcmljdWx1bS5jb20iLCJpYXQiOjE2NTA5MTQ2NzksImV4cCI6MTY1MTUxOTQ3OSwiYXpwIjoiSDR1VHJzdjJoMGlEVGlTMDR2NmVGWmNpdTNLMGJvWnIiLCJndHkiOiJjbGllbnQtY3JlZGVudGlhbHMifQ.X6i4rNvo21ycR4l8Vbdovkyhc7lBYEdQ3zOzexOLO00XfcTO08wihZXuCHESCuYWxej39FLgNvoYINIqenRTEJkkyWR08KC_ONDUXpYvmJivFfmMzaBcDv4J9UvPxb0den-LMT-dfbAiwqQGXL1DadAo3nMHuzUFVpeLcJZ4lqQplulLPKuq9Mbsjfe4XC3Y6pm0sBd-0KI_MMYEBKHIw9U_arR_wf7GmcD3R_DZ-kzOwvmknu4VkmhGlAgzrZqq2uHUXzHBUOY99i_P2PktB5Ty7d9yKksSi5fJgQn6yTbRKghu5keZQ5lwaTfuq5PGovSIF-jDwiFHhf0qT3URsw';
```
Comment (zabeehpericulum) [inline example/lib/main.dart:34]: no, please never ever code the token into the SDK. This should be provided by the developer from their own backends. I can go over this with you if you need clarification.

File: example/lib/main.dart [hunk 5]
```diff
@@ -58,7 +61,7 @@ class _MyAppState extends State<MyApp> {
...
+                                  "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1VSkJOVUk0UkRFek9FVTBORGd4UWpVMVJqTTJPVEJEUXpRMFF6bEJRa1F6UWpnd1JETkVSQSJ9.eyJodHRwczovL2luc2lnaHRzLXBlcmljdWx1bS5jb20vdGVuYW50IjoibnVjbGV1c2lzIiwiaXNzIjoiaHR0cHM6Ly9wZXJpY3VsdW0tdGVjaG5vbG9naWVzLWluYy5hdXRoMC5jb20vIiwic3ViIjoiSDR1VHJzdjJoMGlEVGlTMDR2NmVGWmNpdTNLMGJvWnJAY2xpZW50cyIsImF1ZCI6Imh0dHBzOi8vYXBpLmluc2lnaHRzLXBlcmljdWx1bS5jb20iLCJpYXQiOjE2NTA5MTQ2NzksImV4cCI6MTY1MTUxOTQ3OSwiYXpwIjoiSDR1VHJzdjJoMGlEVGlTMDR2NmVGWmNpdTNLMGJvWnIiLCJndHkiOiJjbGllbnQtY3JlZGVudGlhbHMifQ.X6i4rNvo21ycR4l8Vbdovkyhc7lBYEdQ3zOzexOLO00XfcTO08wihZXuCHESCuYWxej39FLgNvoYINIqenRTEJkkyWR08KC_ONDUXpYvmJivFfmMzaBcDv4J9UvPxb0den-LMT-dfbAiwqQGXL1DadAo3nMHuzUFVpeLcJZ4lqQplulLPKuq9Mbsjfe4XC3Y6pm0sBd-0KI_MMYEBKHIw9U_arR_wf7GmcD3R_DZ-kzOwvmknu4VkmhGlAgzrZqq2uHUXzHBUOY99i_P2PktB5Ty7d9yKksSi5fJgQn6yTbRKghu5keZQ5lwaTfuq5PGovSIF-jDwiFHhf0qT3URsw");
```
Comment (zabeehpericulum) [inline example/lib/main.dart:64]: again never hardcode the token in. This should not have been done. I wasn't able to review the code the previous developers worked on so I missed this until now.

File: example/lib/main.dart [hunk 6]
```diff
@@ -105,7 +108,7 @@ class _MyAppState extends State<MyApp> {
...
```
Comment (zabeehpericulum) [inline example/lib/main.dart:111]: remove this please.

File: example/lib/main.dart [hunk 7]
```diff
@@ -127,16 +130,91 @@ class _MyAppState extends State<MyApp> {
...
+                            isLoading = true;
+                          });
+                          var response;
+                          try {
+                            response = await FlutterPericulum.statementAnalytics(
+                                token:
+                                "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1VSkJOVUk0UkRFek9FVTBORGd4UWpVMVJqTTJPVEJEUXpRMFF6bEJRa1F6UWpnd1JETkVSQSJ9.eyJodHRwczovL2luc2lnaHRzLXBlcmljdWx1bS5jb20vdGVuYW50IjoibnVjbGV1c2lzIiwiaXNzIjoiaHR0cHM6Ly9wZXJpY3VsdW0tdGVjaG5vbG9naWVzLWluYy5hdXRoMC5jb20vIiwic3ViIjoiSDR1VHJzdjJoMGlEVGlTMDR2NmVGWmNpdTNLMGJvWnJAY2xpZW50cyIsImF1ZCI6Imh0dHBzOi8vYXBpLmluc2lnaHRzLXBlcmljdWx1bS5jb20iLCJpYXQiOjE2NTA5MTQ2NzksImV4cCI6MTY1MTUxOTQ3OSwiYXpwIjoiSDR1VHJzdjJoMGlEVGlTMDR2NmVGWmNpdTNLMGJvWnIiLCJndHkiOiJjbGllbnQtY3JlZGVudGlhbHMifQ.X6i4rNvo21ycR4l8Vbdovkyhc7lBYEdQ3zOzexOLO00XfcTO08wihZXuCHESCuYWxej39FLgNvoYINIqenRTEJkkyWR08KC_ONDUXpYvmJivFfmMzaBcDv4J9UvPxb0den-LMT-dfbAiwqQGXL1DadAo3nMHuzUFVpeLcJZ4lqQplulLPKuq9Mbsjfe4XC3Y6pm0sBd-0KI_MMYEBKHIw9U_arR_wf7GmcD3R_DZ-kzOwvmknu4VkmhGlAgzrZqq2uHUXzHBUOY99i_P2PktB5Ty7d9yKksSi5fJgQn6yTbRKghu5keZQ5lwaTfuq5PGovSIF-jDwiFHhf0qT3URsw")
```
Comment (zabeehpericulum) [inline example/lib/main.dart:148]: remove the token please out of source control.

File: example/lib/main.dart [hunk 8]
```diff
@@ -127,16 +130,91 @@ class _MyAppState extends State<MyApp> {
...
+                            isLoading = true;
+                          });
+                          var response;
+                          try {
+                            response =
+                                await FlutterPericulum.statementTransaction(token: "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Ik1VSkJOVUk0UkRFek9FVTBORGd4UWpVMVJqTTJPVEJEUXpRMFF6bEJRa1F6UWpnd1JETkVSQSJ9.eyJodHRwczovL2luc2lnaHRzLXBlcmljdWx1bS5jb20vdGVuYW50IjoibnVjbGV1c2lzIiwiaXNzIjoiaHR0cHM6Ly9wZXJpY3VsdW0tdGVjaG5vbG9naWVzLWluYy5hdXRoMC5jb20vIiwic3ViIjoiSDR1VHJzdjJoMGlEVGlTMDR2NmVGWmNpdTNLMGJvWnJAY2xpZW50cyIsImF1ZCI6Imh0dHBzOi8vYXBpLmluc2lnaHRzLXBlcmljdWx1bS5jb20iLCJpYXQiOjE2NTA5MTQ2NzksImV4cCI6MTY1MTUxOTQ3OSwiYXpwIjoiSDR1VHJzdjJoMGlEVGlTMDR2NmVGWmNpdTNLMGJvWnIiLCJndHkiOiJjbGllbnQtY3JlZGVudGlhbHMifQ.X6i4rNvo21ycR4l8Vbdovkyhc7lBYEdQ3zOzexOLO00XfcTO08wihZXuCHESCuYWxej39FLgNvoYINIqenRTEJkkyWR08KC_ONDUXpYvmJivFfmMzaBcDv4J9UvPxb0den-LMT-dfbAiwqQGXL1DadAo3nMHuzUFVpeLcJZ4lqQplulLPKuq9Mbsjfe4XC3Y6pm0sBd-0KI_MMYEBKHIw9U_arR_wf7GmcD3R_DZ-kzOwvmknu4VkmhGlAgzrZqq2uHUXzHBUOY99i_P2PktB5Ty7d9yKksSi5fJgQn6yTbRKghu5keZQ5lwaTfuq5PGovSIF-jDwiFHhf0qT3URsw", key: "120").then((value) => {
```
Comment (zabeehpericulum) [inline example/lib/main.dart:185]: please remove the token out of source control.

File: lib/models/ExistingStatment.dart [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+import 'dart:convert';
+
+List<Statement> statementFromJson(String str) =>
+    List<Statement>.from(json.decode(str).map((x) => Statement.fromJson(x)));
+
+String statementToJson(List<Statement> data) =>
+    json.encode(List<dynamic>.from(data.map((x) => x.toJson())));
+
+class Statement {
```
Comment (zabeehpericulum) [inline lib/models/ExistingStatment.dart:13]: this model is missing many many different things. Please refer to this: https://www.periculum.io/documentation/insights/#responseanalytics

Comment (zabeehpericulum) [reply lib/models/ExistingStatment.dart:13]: I would also rename this file to just Statement.

File: lib/flutter_periculum.dart [hunk 10]
```diff
@@ -68,4 +69,46 @@ class FlutterPericulum {
 
     return jsonEncode(myresponse).toString();
   }
+
+  static Future<dynamic> statementAnalytics({required String token}) async {
```
Comment (zabeehpericulum) [inline lib/flutter_periculum.dart:73]: there should be no dynamics in this file. We should have models for everything. In this call, make it return an Analytics model.

File: lib/flutter_periculum.dart [hunk 11]
```diff
@@ -68,4 +69,46 @@ class FlutterPericulum {
...
+      myResponse = {
+        "status": false,
+        "msg": result,
+      };
+    }
+    //return an object of {"status": true, "data": (response)}
+    return myResponse;
+  }
+
+  static Future<dynamic> statementTransaction(
```
Comment (zabeehpericulum) [inline lib/flutter_periculum.dart:92]: This should return a Transactions array model.

File: lib/flutter_periculum.dart [hunk 12]
```diff
@@ -68,4 +69,46 @@ class FlutterPericulum {
 
     return jsonEncode(myresponse).toString();
   }
+
```
Comment (zabeehpericulum) [inline lib/flutter_periculum.dart:72]: Also please update the affordability call to also return an Affordability model.

File: android/src/main/kotlin/com/periculum/flutter_periculum/models/requests/PericulumData.kt [hunk 13]
```diff
@@ -113,3 +113,18 @@ data class SMS(
     val announcements_subtype: Int = 0,
 )
 
+data class Statement (
```
Comment (zabeehpericulum) [inline android/src/main/kotlin/com/periculum/flutter_periculum/models/requests/PericulumData.kt:116]: I don't understand this. What is this for?

File: android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt [hunk 14]
```diff
@@ -142,7 +143,60 @@ class FlutterPericulumPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
...
+            println("getStatementTransaction..")
+            val args = call.arguments as HashMap<String, Any>
+            if (args != null){
+                var token = args.get("token")
+                var key = args.get("statementKey")
+                val endpoint = "/statements"
+                val client = OkHttpClient()
+                var request = Request.Builder()
+                    .addHeader("Authorization", "Bearer $token")
+                    .url("$BASE_URL$endpoint/$key/transactions/")
```
Comment (zabeehpericulum) [inline android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt:182]: I don't think you need the backslash after `...transactions`.

File: android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt [hunk 15]
```diff
@@ -142,7 +143,60 @@ class FlutterPericulumPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
...
+        } else if (call.method == "getStatementAnalytics"){
+            println("getStatementAnalytics..")
+            val args = call.arguments as HashMap<String, Any>
+            if (args != null){
+                val token = args.get("token")
+                val endpoint = "statement"
+                val client = OkHttpClient()
+                var request = Request.Builder()
+                        .addHeader("Authorization", "Bearer $token")
+                        .url("$BASE_URL/statements/")
```
Comment (zabeehpericulum) [inline android/src/main/kotlin/com/periculum/flutter_periculum/FlutterPericulumPlugin.kt:155]: please remove the backslash at the end.

**Review by zabeehpericulum — CHANGES_REQUESTED** [review]

File: example/lib/main.dart (hunk 5 above)
Comment (SteveOye) [reply example/lib/main.dart:64]: Okay.. the previous dev used it that way.. So, I will take them off..

File: example/lib/main.dart (hunk 4 above)
Comment (SteveOye) [reply example/lib/main.dart:34]: Yea.. Just use that for test.. Will take it out right away..

Comment (SteveOye) [reply example/lib/main.dart:34]: That has been made provision for. The token here was only for the test purpose

**Pull request closed without merging**